internal/agentapi/client.go

```diff
@@ -0,0 +1,69 @@
+package agentapi
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/buildkite/agent/v3/internal/socket"
+)
+
+const lockAPIPrefix = "http://agent/api/leader/v0/lock/"
+
+// Client is a client for the agent API socket.
+type Client struct {
+	sc *socket.Client
+}
+
+// NewClient creates a new Client using the socket at a given path. The context
+// is used for an internal check that the socket can be dialled.
+func NewClient(ctx context.Context, path string) (*Client, error) {
+	// The API is unauthenticated, hence the empty token.
+	sc, err := socket.NewClient(ctx, path, "")
+	if err != nil {
+		return nil, err
+	}
+	return &Client{sc: sc}, nil
+}
+
+// Ping pings the server. It returns a non-nil error if the ping fails, or the
+// response timestamp is more than 100 milliseconds different to time.Now.
+func (c *Client) Ping(ctx context.Context) error {
+	var resp PingResponse
+	if err := c.sc.Do(ctx, "GET", "http://agent/api/leader/v0/ping", nil, &resp); err != nil {
+		return err
+	}
+	if time.Since(resp.Now) > 100*time.Millisecond {
+		return fmt.Errorf("ping timestamp %v too old", resp.Now)
+	}
```

Comment:
i think this timestamp checking is a good idea in general, but is there a specific failure case that it'll catch? my understanding is that if the ping takes >100ms, it'll cause a leadership coup, is that right?

Reply:
Not really, no. It feels like if the agent is up and running, it ought to be able to respond to a local ping that quickly. Possibly it could be slow if the system as a whole is so overloaded, but then churning the leader won't help.

[Side note: I like 100ms because it's simultaneously "blink of an eye" for a human and "a long time" for a computer.]

Reply:
sweet 👍 i think it's still worthwhile to keep this around, i was just interested to know if there was something i'm missing